index/mem/index.go

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"crypto/md5"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+const initialSize = 1024 * 1024
+
+// TODO(prateek): investigate impact of native heap
+type memIndex struct {
+	opts   Options
+	hashFn doc.HashFn
+
+	docIDGen *atomic.Uint32
+
+	// hash(doc.ID)  -> internal docID
+	idsLock sync.RWMutex
+	idsMap  map[doc.Hash]index.DocID
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[index.DocID]document
+
+	// fieldName -> (Term -> postingID)
+	fieldLock sync.RWMutex
+	fieldMap  map[string]*termsDict
+
+	// postingID -> bitmap of docIDs
+	postingsList postingsList
+}
+
+type document struct {
+	doc.Document
+	docID      index.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(opts Options) Index {
+	hashFn := func(i doc.ID) doc.Hash {
+		return md5.Sum([]byte(i))
+	}
+
+	return &memIndex{
+		opts:   opts,
+		hashFn: hashFn,
+
+		docIDGen: atomic.NewUint32(0),
+
+		idsMap:       make(map[doc.Hash]index.DocID, initialSize),
+		documents:    make(map[index.DocID]document, initialSize),
+		fieldMap:     make(map[string]*termsDict, initialSize),
+		postingsList: newPostingsList(initialSize),
+	}
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *memIndex) Insert(d doc.Document) error {
+	// early terminate if id is already present
+	docHash := i.hashFn(d.ID)
+	i.idsLock.RLock()
+	_, ok := i.idsMap[docHash]
+	i.idsLock.RUnlock()
+	if ok {
+		return nil
```

Comment:
I think we should return an error here. Although the ID is the same, the contents of the document may be different from what is already stored in the index and the caller would have no way of knowing that we didn't store the contents of the document if they did change.

Reply:
Agreed. Will do.

Reply:
done at the Index level (instead of at the Segment level)